pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,318 @@
+	hostname = fmt.Sprintf("%s.%s.svc.cluster.local", name, namespace)
+	cluster = model.BuildSubsetKey(model.TrafficDirectionOutbound, "", host.Name(hostname), port)
+	return hostname, cluster, nil
+}
+
+func parsePort(port uint32) (int, error) {
+	if 1 <= port && port <= 65535 {
+		return int(port), nil
+	}
+	return 0, fmt.Errorf("port must be in the range [1, 65535], found: %d", port)
+}
+
+func parseStatusOnError(status string) (*envoytypev3.HttpStatus, error) {
+	if status == "" {
+		return nil, nil
+	}
+	code, err := strconv.ParseInt(status, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("invalid statusOnError %s: %v", status, err)
+	}
+	if _, found := envoytypev3.StatusCode_name[int32(code)]; !found {
+		return nil, fmt.Errorf("unsupported statusOnError %s, supported values: %v", status, envoytypev3.StatusCode_name)
+	}
+	return &envoytypev3.HttpStatus{Code: envoytypev3.StatusCode(code)}, nil
+}
+
+func generateHTTPConfig(hostname, cluster string, status *envoytypev3.HttpStatus,
+	config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationHttpProvider) *builtExtAuthz {
+	service := &extauthzhttp.HttpService{
+		PathPrefix: config.PathPrefix,
+		ServerUri: &envoy_config_core_v3.HttpUri{
+			// Timeout is required. Use a large value as a placeholder and so that the timeout in DestinationRule (should
+			// be much smaller) can be used to control the real timeout.
+			Timeout: &duration.Duration{Seconds: 600},
+			// Uri is required but actually not used in the ext_authz filter.
+			Uri: fmt.Sprintf("http://%s", hostname),
+			HttpUpstreamType: &envoy_config_core_v3.HttpUri_Cluster{
+				Cluster: cluster,
+			},
+		},
+	}
+	if headers := generateHeaders(config.IncludeHeadersInCheck); headers != nil {
+		service.AuthorizationRequest = &extauthzhttp.AuthorizationRequest{
+			AllowedHeaders: headers,
+		}
+	}
+	if len(config.HeadersToUpstreamOnAllow) > 0 || len(config.HeadersToDownstreamOnDeny) > 0 {
+		service.AuthorizationResponse = &extauthzhttp.AuthorizationResponse{
+			AllowedUpstreamHeaders: generateHeaders(config.HeadersToUpstreamOnAllow),
+			AllowedClientHeaders:   generateHeaders(config.HeadersToDownstreamOnDeny),
+		}
+	}
+	http := &extauthzhttp.ExtAuthz{
+		StatusOnError:       status,
+		FailureModeAllow:    config.FailOpen,
+		TransportApiVersion: envoy_config_core_v3.ApiVersion_V3,
+		Services: &extauthzhttp.ExtAuthz_HttpService{
+			HttpService: service,
+		},
+		FilterEnabledMetadata: generateFilterMatcher(authzmodel.RBACHTTPFilterName),
+	}
+	return &builtExtAuthz{http: http}
+}
+
+func generateGRPCConfig(cluster string, failOpen bool, status *envoytypev3.HttpStatus) *builtExtAuthz {
+	// The cluster includes the character `|` that is invalid in gRPC authority header and will cause the connection
+	// rejected in the server side, replace it with a valid character and set in authority otherwise ext_authz will
+	// use the cluster name as default authority.
+	authority := strings.ReplaceAll(cluster, "|", "-")
```

Comment:
We already have a header compatible, BuildDNSSrvSubsetKey. Should we use this?

Reply:
probably not, I think this is only used in `sni` but not the actual cluster name?

Reply:
Why not? I don't want to invent a new format. BuildDNSSrvSubsetKey is meant for _exactly_ this use case.

Reply:
Do you mean to change the networking code to use BuildDNSSrvSubsetKey for the cluster name in the first place? If that is okay (not sure if it has any significant side effect),  that could be done as a separate refactor in the networking core code.

Here we are only referring to the already generated cluster name that still uses the `|` format.

Reply:
No, I mean instead of replacing `|` with `-` replace it with `_.` which is already the standardized format

Reply:
Oh, I misunderstood your original comment, will do in a follow PR.